app/src/main/kotlin/compiler/intermediate/ControlFlow.kt

```diff
@@ -502,44 +536,75 @@ object ControlFlow {
-
-            fun processStatementBlock(block: StatementBlock, returnCleanup: (() -> Unit)? = null) { // returnCleanup asserts that all returning nodes are in "last"
-                fun addCfg(cfg: ControlFlowGraph): IFTNode? {
-                    val entry = cfg.entryTreeRoot
-                    if (entry != null) {
-                        for (node in last) {
-                            cfgBuilder.addLink(node, entry)
-                        }
 
-                        last = cfg.finalTreeRoots
+            fun addCfg(cfg: ControlFlowGraph): IFTNode? {
+                val entry = cfg.entryTreeRoot
+                if (entry != null) {
+                    for (node in last) {
+                        cfgBuilder.addLink(node, entry)
                     }
-                    cfgBuilder.addAllFrom(cfg)
-                    return entry
+
+                    last = cfg.finalTreeRoots
                 }
+                cfgBuilder.addAllFrom(cfg)
+                return entry
+            }
+
+            fun addNode(node: IFTNode) {
+                addCfg(ControlFlowGraphBuilder().addSingleTree(node).build())
+            }
+
+            fun addDetachedNode(node: IFTNode) {
+                cfgBuilder.addAllFrom(ControlFlowGraphBuilder().addSingleTree(node).build(), false)
+            }
 
-                fun addNode(node: IFTNode) {
-                    addCfg(ControlFlowGraphBuilder().addSingleTree(node).build())
+            fun Variable.generateDestructor(): ControlFlowGraph = when (type) {
+                is Type.Array -> ControlFlowGraphBuilder()
+                    .mergeUnconditionally(arrayMemoryManagement.genRefCountDecrement(createReadFromVariable(this, function), type))
+                    .addSingleTree(createWriteToVariable(IFTNode.Const(0), this, function))
+                    .build()
+                else -> ControlFlowGraphBuilder().build()
+            }
+
+            fun processStatementBlock(block: StatementBlock, returnTreeHead: EscapeTreeHead, breakTreeHead: EscapeTreeHead?, continueTreeHead: EscapeTreeHead?) {
+                val destructorsStack = Stack<ControlFlowGraph>()
+
+                fun addToEscapeTree(value: ControlFlowGraph, treeHead: EscapeTreeHead?) {
+                    if (treeHead == null || value.entryTreeRoot == null)
+                        return
+                    cfgBuilder.addAllFrom(value)
+                    value.finalTreeRoots.forEach { cfgBuilder.addLink(it, treeHead.value, false) }
+                    treeHead.value = value.entryTreeRoot
+                }
+
+                fun addToDestructionStructures(variable: Variable) {
+                    val genDestructor = { variable.generateDestructor() }
+                    destructorsStack.add(genDestructor())
+                    addToEscapeTree(genDestructor(), returnTreeHead)
+                    addToEscapeTree(genDestructor(), breakTreeHead)
+                    addToEscapeTree(genDestructor(), continueTreeHead)
```

Comment:
Why are we adding four different constructors for one variable? Can one variable be destructed multiple times, or are we ensuring that only of these will get called?

Reply:
I'm pretty sure you meant 'destructors'. And your second guess was correct, only one of them will be called (at least before variable is initialized again):

`destructorsStack` represents variables being destroyed because block they are defined in ends (destructors are inserted at the end of proper block).

Escape trees represent variables being destroyed because of `return`, `break` or `continue`. There is one returnEscapeTree for entire function and one break-/continueEscapeTree for each loop.

These four can't be merged (or at least I don't know how to do it in reasonable way), because even if, for example, `continue` and `break` would destroy the same variables, we need to end up in different place in code after that.